### DISCUSSION TITLE
Simultaneous use of an MCP2515 and CMT2300A

### DISCUSSION THREAD

**Discussion by Manos1966** [skippermeister/OpenDTU-OnBattery#11] [General]
Hi @skippermeister 

Bezueglich des Wiki Eintrags:
https://github.com/skippermeister/OpenDTU-OnBattery/wiki#15-simultaneous-use-of-an-mcp2515-and-cmt2300a

> This means that the CMT2300A module can be used alongside the MCP2515 for the Huwei or Meanwell power supply (Huawei is not yet tested).

Soll ich es mit Huawei testen?

**Comment by skippermeister** [comment]
Hi @Manos1966 
Du kannst es gerne Test. Ich habe selbst kein Huawei Netzteil. Du musst Dir aber ein eigenes Binär File erstellen und dazu in der Platformio.ini das Meanwell Modul deaktivieren und das Huawei Modul aktivieren

**Reply by Manos1966** [reply]
> Du musst Dir aber ein eigenes Binär File erstellen und dazu in der Platformio.ini das Meanwell Modul deaktivieren und das Huawei Modul aktivieren

🙈 🙈 🙈 🙈 🙈 🤦‍♂️

**Reply by skippermeister** [reply]
Was für ein ESP32 Modul hast Du zum Testen. Ich kann Dir ein Binär File erstellen und hochladen.

**Reply by Manos1966** [reply]
Was moechten wir getestet haben? 🤔 

Eigentlich nur deine Version mit MC2515 (Huawei CANBUS)
Dann ESP32-S3

Soll gleichzeitig ein RS485 Typ A oder Type B angeschlossen sein? (ohne waere es fuer mich einfacher)

**Reply by skippermeister** [reply]
@Manos1966 
eine neues release mit Huawei Ladegerät Unterstützung ist nun hochgeladen. Du findest jetzt Bin-Dateien für Huawei und für Meanwell Ladegeräte.
Es gibt CAN Bus Unterstützung für MCP2515, I2C to MCP2515 und interner CAN0 Bus.
Du selektierst den jeweiligen CAN Bus Treiber über das PinMapping File. Siehe dazu in meinem Wiki.

RS485 A oder B wird ja für die Batterie benötigt.

Zunächst nur das Huawei Ladegerät über MCP2515 testen.

Schick mir dazu Log Dateien und einen Screen Shot nachdem hoffentlich das Huawei Ladegerät erkannt wurde.

**Reply by sentinel1111** [reply]
Hallo,

Ich habe heute das Huawei Ladegerät mit MCP2515 und CMT2300 getestet.Ladegerät wird erkannt.120 Ohm muß gebrückt sein.
Ich habe bei mir die MCP2515 Schaltung auf der Platine,sollte aber kein unterschied machen.
I2C Version muß ich noch testen,da hatte ich letztens noch Probleme.

![grafik](https://github.com/user-attachments/assets/919f0841-3873-4969-8a65-68db83a9703f)

![grafik](https://github.com/user-attachments/assets/96e9064b-3908-4d58-9b1a-1f1fbbd57149)

[OdTUputty.txt](https://github.com/user-attachments/files/16922470/OdTUputty.txt)

**Reply by Manos1966** [reply]
Hi @skippermeister 
bei mir wurde nicht erkannt.

Ich habe opendtu-onbattery-generic_esp32s3_huawei.factory.bin
geflashed.

Die gleiche Hardware, mit Andreas Firmware funktioniert einwandfrei.

Meine Vermutung: Du hast keine Option, die Taktfrequenz des MC2515 zwischen 8MHz und 16MHz auszuwaehlen und du hast eine 8MHz Version (waehrend meiner 16MHz Frequenz hat)
Hier ist ein Bild von Andreas Setup:
![not 04](https://github.com/user-attachments/assets/9a3c5069-0c40-4a44-a9ca-82096309fb9e)


---------------------------------
Network lost connection
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0xc (RTC_SW_CPU_RST),boot:0xb (SPI_FAST_FLASH_BOOT)
Saved PC:0x421351f6
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x4bc
load:0x403c9700,len:0xbd8
load:0x403cc700,len:0x2a0c
entry 0x403c98d0
E (374) esp_core_dump_flash: No core dump partition found!
E (375) esp_core_dump_flash: No core dump partition found!

Starting OpenDTU-onBattery
Initialize FS... done
Reading configuration... [Configuration]read migrate PowerLimiter Inverter ID:00 to Serial No: 0
done
Reading PinMapping... found valid mapping 04 ESP32-S3 NO-NRF+NO-CMT+CAN+HUAWEI ... done
[SerialPortManager] HW UART port 0 now in use by 'Serial Console'
Initialize Network... done
Setting Hostname... Configuring WiFi STA using new credentials... done
Initialize NTP... done
Initialize SunPosition... done.
Initialize MqTT... done
Initialize WebApi... done
Initialize Display... none configured. 
Initialize LEDs... done
Check for default DTU serial... done
initialize Hoymiles interface... Invalid pin config
initialize Datastore... done
Initialize VE.Direct interface...
[VictronMppt]initController RS232 port rx = -1, tx = -1, instance = 1, not configued
[VictronMppt]initController RS232 port rx = -1, tx = -1, instance = 2, not configued
[VictronMppt]initController RS232 port rx = -1, tx = -1, instance = 3, not configued
done
Initialize Power Limiter... no MosFET pins configured to control DC power of inverter 0 channel 0
no Inverter for Powerlimiter configured
Initialize Dynamic Zero Grid Export Limiter... done
Invalid pin config
[HuaweiCanClass]::updateSettings Error Initializing Huawei CAN communication...
Battery provider not enabled
Free heap: 177616
Switch to WiFi mode
Setting Hostname... done
Configuring WiFi STA using new credentials... done
Configuring WiFi STA DHCP IP... done
[PowerLimiter]announceStatus: disabled by configuration
WiFi connected
WiFi got ip: 192.168.188.198
Network connected
Admin AP remaining seconds: 10 / 180
Admin AP remaining seconds: 20 / 180
Admin AP remaining seconds: 30 / 180
Websocket: [/livedata][1] connect
Websocket: [/vedirectlivedata][1] connect
---------------------------------

**Reply by Manos1966** [reply]
@sentinel1111 was steht auf dem Crystal oben Rechts auf deine Platine? 8.000 oder 16.0 ?

**Reply by skippermeister** [reply]
@Manos1966 
Dein PinMapping ist fehlerhaft. Mein PinMapping sieht anders aus, als das von der original openDTU oB Software. Bitte schaue in meinem Wiki nach.

Die Huawei Initialisierung bricht mit diesem Fehler ab:
Invalid pin config
[HuaweiCanClass]::updateSettings Error Initializing Huawei CAN communication...

Die Frequenz des Quarzes kannst Du im Setup des AC Chargers einstellen, sofern das PinMapping für den MCP2515 korrekt erkannt wurde. 
Falls im Setup Menü nicht angezeigt wird:
Aktiviere Huawei R4850G2 an MCP2515 CAN Bus Interface 
dann ist Dein PinMapping fehlerhaft.

**Reply by sentinel1111** [reply]
@Manos1966 

Es sind 16MHz. Einstellbar im Setup des AC Chargers.

Pimapp sieht so aus:

"charger": {
            "mcp2515_irq": -1,
            "mcp2515_clk": -1,
            "mcp2515_mosi": -1,
            "mcp2515_miso": -1,
            "mcp2515_cs": -1,
            "power": -1

**Reply by Manos1966** [reply]
@skippermeister 

Mist! 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ 🤦‍♂️ 

Sorry Jungs, ich  bin nicht drauf gekommen 🤦‍♂️ 

Dann ist alles PERFEKT! 🥂 

Uebrigens nur zur Info: Es gibt zur Zeit Huawei R4850G2 Netzteile zum "anständigen" Preis (70EUR) bei eBay-Kleinanzeigen.
Ich habe gerade eins bestellt,  ich werde nächste Woche sagen können wie gut es funktioniert.. 
Adaptor gibt es hier:
![Huawei R4850 Connector](https://github.com/user-attachments/assets/c8797329-4239-43bd-8e38-2fbbdc873f9b)
